apm-agent-plugins/apm-jakarta-websocket-plugin/src/main/java/co/elastic/apm/agent/websocket/BaseServerEndpointInstrumentation.java

```diff
@@ -0,0 +1,116 @@
+import net.bytebuddy.matcher.ElementMatchers;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.classLoaderCanLoadClass;
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.isInAnyPackage;
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+import static net.bytebuddy.matcher.ElementMatchers.isAnnotatedWith;
+import static net.bytebuddy.matcher.ElementMatchers.isBootstrapClassLoader;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.namedOneOf;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+
+public abstract class BaseServerEndpointInstrumentation extends TracerAwareInstrumentation {
+
+    private final Collection<String> applicationPackages;
+
+    public BaseServerEndpointInstrumentation(ElasticApmTracer tracer) {
+        applicationPackages = tracer.getConfig(StacktraceConfiguration.class).getApplicationPackages();
+    }
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        return not(isBootstrapClassLoader())
+            .and(classLoaderCanLoadClass(getServerEndpointClassName()));
+    }
+
+    @Override
+    public ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+        return isInAnyPackage(applicationPackages, ElementMatchers.<NamedElement>any());
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return isAnnotatedWith(named(getServerEndpointClassName()));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return isAnnotatedWith(
+            namedOneOf("javax.websocket.OnOpen", "jakarta.websocket.OnOpen")
+                .or(namedOneOf("javax.websocket.OnMessage", "jakarta.websocket.OnMessage"))
+                .or(namedOneOf("javax.websocket.OnError", "jakarta.websocket.OnError"))
+                .or(namedOneOf("javax.websocket.OnClose", "jakarta.websocket.OnClose")));
+    }
+
+    protected abstract String getServerEndpointClassName();
+
+    protected static class BaseAdvice {
+
+        protected static Object startTransactionOrSetTransactionName(String signature, String frameworkName, @Nullable String frameworkVersion) {
+            Transaction currentTransaction = tracer.currentTransaction();
+            if (currentTransaction == null) {
+                Transaction rootTransaction = tracer.startRootTransaction(Thread.currentThread().getContextClassLoader());
+                if (rootTransaction != null) {
+                    setTransactionName(rootTransaction, signature, frameworkName, frameworkVersion);
+                    return rootTransaction.activate();
+                }
+            } else {
+                setTransactionName(currentTransaction, signature, frameworkName, frameworkVersion);
```

Comment:
When does this case happen? Is it when one websocket-annotated method calls another one?
Maybe unlikely, but what if such a method is called from a completely unrelated transaction? Will that transaction be renamed?
Could we just return null in this case and avoid renaming the transaction? Or would we miss an important case if we did that?

Reply:
The initial (upgrade) request of a WebSocket connection is a pain http request. Usually this request will be handled by a servlet container. Thus, when the `OnOpen ` method is called the servlet plugin already started a transaction. All other WebSocket request/messages won't be handled by the servlet container anymore.

This is similar to the JAX-RS plugin, which also sets the transaction name for the transaction started by the servlet container.